### PR TITLE
Use new joint handles in all controllers

### LINKED ISSUE #89: Refactor controllers to use the new joint handles
See https://github.com/ros-controls/ros2_control/issues/124.

List of controllers to refactor:
- [x] joint_state_controller
- [x] joint_trajectory_controller
- [x] diff_drive_controller

### PR DESCRIPTION
**Pull request by jordan-palacios**
Refactor joint_state_controller, joint_trajectory_controller and diff_drive_controller to use the new joint handles.

Depends on https://github.com/ros-controls/ros2_control/pull/125 and https://github.com/ros-controls/ros2_control/pull/136.

Closes #89.

### CODE REVIEW

File: joint_state_controller/src/joint_state_controller.cpp [hunk 1]
```diff
@@ -121,17 +178,30 @@ JointStateController::update()
...
+    auto update_value =
+      [](double & msg_data, const std::shared_ptr<hardware_interface::JointHandle> & joint_handle)
+      {
+        if (joint_handle) {
+          msg_data = joint_handle->get_value();
+        }
+      };
+
+    /// @todo somehow don't use hardcoded indexes directly
+    update_value(joint_state_msg_.position[joint_index], joint_handles_[joint_index][0]);
```
Comment (bmagyar) [inline joint_state_controller/src/joint_state_controller.cpp:193]: could you at least make these hardcoded indices named by putting htem in an ALL_CAPS_CONST please?

**Review by bmagyar — CHANGES_REQUESTED** [review]

File: joint_state_controller/src/joint_state_controller.cpp (hunk 1 above)
Comment (jordan-palacios) [reply joint_state_controller/src/joint_state_controller.cpp]: Done.

**Comment by destogl** [comment]
I added myself to review. I hope this is OK. I will do it at latest tomorrow.

**Comment by jordan-palacios** [comment]
> I added myself to review. I hope this is OK. I will do it at latest tomorrow.

Sure. I only ask this is not delayed further. It's currently blocking a bunch of PRs related to the porting of controllers from ROS1 that are pretty much ready. See https://github.com/ros-controls/ros2_controllers/issues/78, https://github.com/ros-controls/ros2_controllers/issues/50, https://github.com/ros-controls/ros2_controllers/issues/51 and https://github.com/ros-controls/ros2_controllers/issues/52.

**Review by bmagyar — APPROVED** [review]
Looking good from my end

**Review by destogl — APPROVED** [review]
It looks excellent!

Thanks for your effort and waiting for my review.

Just two small comments, nothing critical. Maybe there are more places where headers are missing, but if linters are happy I am to :)

File: joint_trajectory_controller/src/joint_trajectory_controller.cpp [hunk 2]
```diff
@@ -26,8 +26,6 @@
 #include "angles/angles.h"
 #include "builtin_interfaces/msg/duration.hpp"
 #include "builtin_interfaces/msg/time.hpp"
-#include "hardware_interface/joint_command_handle.hpp"
```
[the remark this answers is not in the document]
Comment (jordan-palacios) [reply joint_trajectory_controller/src/joint_trajectory_controller.cpp:29]: It comes free from already including `robot_hardware.hpp`. Doesn't hurt to add it too though.

[the remark this answers is not in the document]
Comment (jordan-palacios) [reply joint_trajectory_controller/src/joint_trajectory_controller.cpp:29]: Done.

File: joint_state_controller/src/joint_state_controller.cpp [hunk 3]
```diff
@@ -107,6 +83,90 @@ JointStateController::on_deactivate(const rclcpp_lifecycle::State & /*previous_s
...
+        joint_interface_names.begin(), joint_interface_names.end(),
+        std::back_inserter(interface_names), [&](const auto & value) {
+          return std::find(
+            interface_names.begin(),
+            interface_names.end(), value) == interface_names.end();
+        });
+
+      std::vector<std::shared_ptr<hardware_interface::JointHandle>> joint_interface_handles;
+      for (const auto & interface_name : interface_names) {
+        /// @todo is there a better way of skipping command interfaces?
```
[the remark this answers is not in the document]
Comment (bmagyar) [reply joint_state_controller/src/joint_state_controller.cpp:118]: https://github.com/ros-controls/ros2_controllers/issues/96

**Pull request merged**